### PR TITLE
Implement upserts

### PR DESCRIPTION
**Pull request by lilsweetcaligula**
Merge BLOCKED by: https://github.com/senecajs/seneca-store-test/pull/59
`seneca-store-test v3.2.0` must be published before tests in this PR may pass.

This PR implements upserts for the mongo store plugin.

### CODE REVIEW

File: mongo-store.js [hunk 1]
```diff
@@ -117,6 +117,10 @@ module.exports = function (opts) {
   function error(args, err, cb) {
     if (err) {
       seneca.log.error('entity', err, { store: name })
+
```
Comment (rjrodger) [inline mongo-store.js:120]: why?

File: mongo-store.js [hunk 2]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
     },
 
     save: function (args, cb) {
-      var ent = args.ent
+      const is_update = Boolean(args.ent.id)
```
Comment (rjrodger) [inline mongo-store.js:188]: Please use convention `null ==` | `null !=`:

`const is_update = null != args.ent.id`

File: mongo-store.js [hunk 3]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
 
-      var update = !!ent.id
+      if (is_update) {
+        return updateExisting(args, cb)
```
Comment (rjrodger) [inline mongo-store.js:191]: early return +1

File: mongo-store.js [hunk 4]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+      function createAndSave(args, cb) {
+        return upsertIfRequested(args, function (err, upsert) {
```
Comment (rjrodger) [inline mongo-store.js:198]: Same feedback as per mem-store - convert to a boolean flag, then operate

File: mongo-store.js [hunk 5]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+          if (error(args, err, cb)) {
+            return
           }
 
-          entp._id = makeid(id)
+          if (upsert.upsert_requested) {
+            return cb(null, upsert.out)
```
Comment (rjrodger) [inline mongo-store.js:204]: replace name `cb` with `done` - this is the modern convention

File: mongo-store.js [hunk 6]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+          }
+
+          return createNew(args, cb)
```
Comment (rjrodger) [inline mongo-store.js:207]: replace `args` with `msg` - again updating to latest convention

File: mongo-store.js [hunk 7]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+        })
+
+
+        function upsertIfRequested(args, cb) {
+          const query_for_save = args.q
+
+          if (Array.isArray(query_for_save.upsert$)) {
```
Comment (rjrodger) [inline mongo-store.js:214]: convert to boolean, perhaps

`const has_upsert = Array.isArray(query_for_save.upsert$)`

File: mongo-store.js [hunk 8]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+            const upsert_on = query_for_save.upsert$
+            const public_entdata = args.ent.data$(false)
+
+            if (upsert_on.length > 0 && upsert_on.every(p => p in public_entdata)) {
```
Comment (rjrodger) [inline mongo-store.js:218]: why the need to check if the upsert prop is in the ent data?
if this fails it's a no-op and there's no feedback
if we pass through "bad" upsert props, mongo should complain, which is what we want

File: mongo-store.js [hunk 9]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+              const filter_by = upsert_on
```
Comment (rjrodger) [inline mongo-store.js:219]: I love a good reduction! +1

File: mongo-store.js [hunk 10]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+                .reduce((acc, field) => {
+                  acc[field] = args.ent[field]
+                  return acc
+                }, {})
+
+              const replacement = (() => {
+                const o = Dot.flatten(Object.assign({}, public_entdata))
```
Comment (rjrodger) [inline mongo-store.js:226]: what is Dot? what does flatten do?

File: mongo-store.js [hunk 11]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+
+
+                const NO_SPECIFIC_ID_REQUESTED = {}
```
Comment (rjrodger) [inline mongo-store.js:229]: why does this replace:
```
          if (undefined !== ent.id$) {
            id = ent.id$
          } else if (opts.generate_id) {
            id = opts.generate_id(ent)
          }

          entp._id = makeid(id)
```

File: mongo-store.js [hunk 12]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+              seneca.log.debug('save/insert', args.ent, desc)
+
+              return cb(null, fent)
+            })
           })
-        } else {
+        }
+
+
+        function tryMakeIdIfRequested(args, { when_no_specific_id_requested }) {
```
Comment (rjrodger) [inline mongo-store.js:320]: OK - strong feedback here - please do not use 'if', 'and', etc in function names. Use the language constructs, and try to keep functions as pure as possible.

Use a `make_id` utility function in this case.

File: mongo-store.js [hunk 13]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+          const ent = args.ent
+
+          if (undefined !== ent.id$) {
```
Comment (rjrodger) [inline mongo-store.js:323]: use `null !=` as it catches both null and undef

File: mongo-store.js [hunk 14]
```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
...
+            return makeid(opts.generate_id(ent))
+          }
+
+          return when_no_specific_id_requested
+        }
+      }
+
+
+      function updateExisting(args, cb) {
+        return getcoll(args, args.ent, function (err, coll) {
```
Comment (rjrodger) [inline mongo-store.js:337]: There is a structural problem here. There should only be one call to getcoll.

File: package.json [hunk 15]
```diff
@@ -1,10 +1,11 @@
 {
   "name": "seneca-mongo-store",
-  "version": "4.0.0",
+  "version": "4.1.0",
```
Comment (rjrodger) [inline package.json:3]: Don't change the version until we are actually publishing

File: package.json [hunk 16]
```diff
@@ -55,7 +56,7 @@
     "coveralls": "^3.1.0",
     "seneca": "^3.22.0",
     "seneca-entity": "^10.0.2",
```
Comment (rjrodger) [inline package.json:58]: freeze the other versions too

File: test/support/helpers.js [hunk 17]
```diff
@@ -0,0 +1,7 @@
+exports.fetchProp = function (o, prop) {
```
Comment (rjrodger) [inline test/support/helpers.js:1]: Please stop using this as we will be converting to TS soon

**Review by rjrodger — CHANGES_REQUESTED** [review]
This store is in production use so this change will need lots of love.

File: mongo-store.js (hunk 10 above)
Comment (lilsweetcaligula) [reply mongo-store.js:226]: @rjrodger 
Dot.flatten basically converts your object to Mongo update directives, which can be then passed to Mongo's `update*` function. E.g.:
```javascript
Dot.flatten({ foo: 'bar' }) // => { $set: { foo: 'bar' } }
```

Source: https://www.npmjs.com/package/mongo-dot-notation

File: test/support/helpers.js (hunk 17 above)
Comment (lilsweetcaligula) [reply test/support/helpers.js:1]: @rjrodger Until we switch to TS, perhaps it would be of benefit to have that extra safety in place?

File: mongo-store.js (hunk 11 above)
Comment (lilsweetcaligula) [reply mongo-store.js:229]: @rjrodger I believe the approach in this PR to be more readable as it hides the details of how it tries to retrieve the id in a small tidy function.

The problem I found with the original code was that it was not clear off the bat that it was intentional that `id` may be undefined. For a moment I actually thought there might have been a bug.

Besides, I strongly believe that `undefined` as a value is not declarative in code. I would personally refrain from using `undefined` in code, where appropriate. The alternative present in this PR is readable and communicates the intention to the reader arguably more clearly.

If you would like to inline the original code back, may I recommend a consensus:
```javascript

let base_id

if (null === ent.id$ || undefined === ent.id$) {
  base_id = ent.id$
} else if (opts.generate_id) {
  base_id = opts.generate_id(ent)
} else {
  base_id = undefined
}

const new_id = makeid(base_id)
entp._id = new_id
```

File: mongo-store.js (hunk 8 above)
Comment (lilsweetcaligula) [reply mongo-store.js:218]: @rjrodger Could you please elaborate? Currently, the code needs to check for that so that if it fails, it knows there can be no successful match, and hence no upsert, - in which case it reverts to creating a new entity.

Comment (lilsweetcaligula) [reply mongo-store.js:218]: Here's a test which serves as an example of the expected behavior:
```javascript
describe('fields in upsert$ are not present in the data$ object', () => {
  beforeEach(() => new Promise(fin => {
    si.make('products')
      .data$({ label: 'a toothbrush', price: '3.40' })
      .save$(fin)
  }))

  it('creates a new entity because it can never match', fin => {
    si.test(fin)

    si.make('products')
      .data$({ label: 'a toothbrush', price: '2.95' })
      .save$({ upsert$: ['label', 'coolness_factor'] }, err => {
        if (err) {
          return fin(err)
        }

        si.make('products').list$({}, (err, products) => {
          if (err) {
            return fin(err)
          }

          expect(products.length).to.equal(2)

          expect(products[0]).to.contain({
            label: 'a toothbrush',
            price: '3.40'
          })

          expect(products[1]).to.contain({
            label: 'a toothbrush',
            price: '2.95'
          })

          return fin()
        })
      })
  })
})
```

File: mongo-store.js [hunk 18]
```diff
@@ -180,79 +181,189 @@ module.exports = function (opts) {
...
+          const public_entdata = msg.ent.data$(false)
+
+          const filter_by = upsert_on
+            .reduce((acc, field) => {
+              acc[field] = msg.ent[field]
+              return acc
+            }, {})
+
+          const replacement = (() => {
+            const o = Dot.flatten(Object.assign({}, public_entdata))
```
Comment (rjrodger) [inline mongo-store.js:232]: We need to remove the dependency on Dot.
Introducing a dependency is a big decision that has consequences for maintenance and security and needs to be examined separately.
Also, we cannot allow the code base to become inconsistent - if Dot is used, then it has to be used in all places where it would make sense.
The introduction of something like Dot would also have to be a separate PR.

File: mongo-store.js [hunk 19]
```diff
@@ -180,79 +181,189 @@ module.exports = function (opts) {
...
+          )
+        }
+
+
+        function createNew(msg, coll, done) {
+          const new_doc = (function () {
+            const public_entdata = msg.ent.data$(false)
+
+
+            const NO_SPECIFIC_ID_REQUESTED = {}
```
Comment (rjrodger) [inline mongo-store.js:270]: Abstract id handing into a utility function called `ent = ensure_id(ent)`. All it needs to do is:
```
          if (undefined !== ent.id$) {
            id = ent.id$
          } else if (opts.generate_id) {
            id = opts.generate_id(ent)
          }

          entp._id = makeid(id)
```

Also be careful of needlessly copying data via Object.assign

File: mongo-store.js [hunk 20]
```diff
@@ -180,79 +181,189 @@ module.exports = function (opts) {
...
+            .reduce((acc, field) => {
+              acc[field] = msg.ent[field]
+              return acc
+            }, {})
+
+          const replacement = (() => {
+            const o = Dot.flatten(Object.assign({}, public_entdata))
+
+
+            const NO_SPECIFIC_ID_REQUESTED = {}
```
Comment (rjrodger) [inline mongo-store.js:235]: this can also be replaced by the `ensure_id` util function as per the comment below

File: mongo-store.js [hunk 21]
```diff
@@ -180,79 +181,189 @@ module.exports = function (opts) {
...
+
+        coll[func](q, set, { upsert: true }, function (err) {
+          if (!error(msg, err, done)) {
+            seneca.log.debug('save/update', ent, desc)
+
+            coll.findOne(q, {}, function (err, entu) {
+              if (!error(msg, err, done)) {
+                var fent = null
+                if (entu) {
+                  entu.id = idstr(entu._id)
```
Comment (rjrodger) [inline mongo-store.js:356]: this looks like common code we can factor out into a util function

File: package.json [hunk 22]
```diff
@@ -45,17 +46,17 @@
   ],
   "license": "MIT",
   "dependencies": {
-    "mongo-dot-notation": "^2.0.0",
-    "mongodb": "^3.5.9"
+    "mongo-dot-notation": "2.0.0",
```
Comment (rjrodger) [inline package.json:49]: remove as per note re Dot above

File: package.json [hunk 23]
```diff
@@ -45,17 +46,17 @@
...
   },
   "devDependencies": {
-    "@hapi/code": "^8.0.1",
-    "@hapi/lab": "^22.0.4",
-    "async": "^3.2.0",
-    "coveralls": "^3.1.0",
-    "seneca": "^3.22.0",
-    "seneca-entity": "^10.0.2",
-    "seneca-store-test": "2.1.0"
+    "@hapi/code": "8.0.1",
```
Comment (rjrodger) [inline package.json:53]: all frozen +1

File: test/mongo.test.js [hunk 24]
```diff
@@ -1,24 +1,23 @@
...
 var Seneca = require('seneca')
 var Async = require('async')
 
 const Lab = require('@hapi/lab')
 const Code = require('@hapi/code')
-const lab = (exports.lab = Lab.script())
 const expect = Code.expect
 
-const describe = lab.describe
+const lab = (exports.lab = Lab.script())
```
Comment (rjrodger) [inline test/mongo.test.js:18]: Only do this if you intend to make this change in the other repos that use Lab

File: test/mongo.test.js [hunk 25]
```diff
@@ -44,24 +43,829 @@ describe('mongo tests', function () {
...
+    before(prepareForRaceConditionTesting)
+
+    after(unprepareAfterRaceConditionTesting)
+
+    Shared.upserttest({
+      seneca: si,
+      script: lab
+    })
+
+    // NOTE: WARNING: The reason we need a unique index on the users.email
```
Comment (rjrodger) [inline test/mongo.test.js:78]: seneca-store-test could define this as a "standard" requirement - that's OK
add a note in the seneca-store-test README

File: mongo-store.js (hunk 19 above)
Comment (lilsweetcaligula) [reply mongo-store.js:270]: Okay, I will extract the code, you have specified in the snippet, into a separate helper called `ensure_id`.

File: mongo-store.js (hunk 18 above)
Comment (lilsweetcaligula) [reply mongo-store.js:232]: The Dot has been listed as a dependency prior to this PR.

File: test/mongo.test.js (hunk 24 above)
Comment (lilsweetcaligula) [reply test/mongo.test.js:18]: If `exports.lab` is removed, I am afraid this happens:

> $ npm run test
> The file: seneca-mongo-store/test/mongo.test.js includes a lab script that is not exported via exports.lab

File: test/mongo.test.js (hunk 25 above)
Comment (lilsweetcaligula) [reply test/mongo.test.js:78]: As requested:
https://github.com/senecajs/seneca-store-test/pull/60/commits/878376b4dbda0e2c15a7d433dff6c66a063aa384
https://github.com/senecajs/seneca-store-test/pull/60/commits/5e9b1c4a907657da743f01e7402af1641ea4bc8d

File: mongo-store.js [hunk 26]
```diff
@@ -180,79 +97,166 @@ module.exports = function (opts) {
...
-            shouldMerge = false
-          }
-          if (opts.merge === false && ent.merge$ !== true) {
-            shouldMerge = false
-          }
 
-          var set = entp
-          var func = 'replaceOne'
+        function createNew(msg, coll, done) {
+          const new_doc = (function () {
```
Comment (rjrodger) [inline mongo-store.js:183]: why is this an IIFE?

**Review by rjrodger — CHANGES_REQUESTED** [review]

File: mongo-store.js (hunk 26 above)
Comment (lilsweetcaligula) [reply mongo-store.js:183]: This is an IIFE for the following purposes:
- grouping code
- if deemed necessary, this approach lends itself to simpler refactoring
- makes it easy to find identical code throughout the code base and DRY it up
- better readability - when you read code that looks like this, it is easier to mentally separate the intention that each group of code tries to convey
- avoids mixing up logic

One might ask: why not just extract it into a separate function? I would argue that extraction of code into a separate function is not always a good idea. In situations when certain logic only appears once throughout the code base - such refactoring would in fact make it harder to understand the code.

File: test/mongo.test.js [hunk 27]
```diff
@@ -44,24 +43,829 @@ describe('mongo tests', function () {
...
   })
 
-  it('extra test', function (done) {
-    extratest(si, done)
+  describe('extra tests', () => {
+    it('extra test', function (done) {
+      extratest(si, done)
+    })
+
+    describe('#save$', () => {
```
[the remark this answers is not in the document]
Comment (lilsweetcaligula) [reply test/mongo.test.js:130]: @rjrodger I have looked into the cause for the bad performance of the tests:
```
$ npm run test-some '#save'

14 tests complete
Test duration: 2478 ms
```

After a brief research, it was revealed to me that the culprit was actually the excessive calls to `Seneca#ready`. After optimization, a considerable improvement in performance can be observed:
```
$ npm run test-some '#save'

14 tests complete
Test duration: 1218 ms
```

This is roughly a 200% performance improvement in the `#save` tests.
